flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/nn/KNN.scala

```diff
@@ -0,0 +1,354 @@
+  *         .setBlocks(5)
+  *         .setDistanceMetric(EuclideanDistanceMetric())
+  *
+  *       knn.fit(trainingDS)
+  *
+  *       val predictionDS: DataSet[(Vector, Array[Vector])] = knn.predict(testingDS)
+  * }}}
+  *
+  * =Parameters=
+  *
+  * - [[org.apache.flink.ml.nn.KNN.K]]
+  * Sets the K which is the number of selected points as neighbors. (Default value: '''5''')
+  *
+  * - [[org.apache.flink.ml.nn.KNN.DistanceMetric]]
+  * Sets the distance metric we use to calculate the distance between two points. If no metric is
+  * specified, then [[org.apache.flink.ml.metrics.distances.EuclideanDistanceMetric]] is used.
+  * (Default value: '''EuclideanDistanceMetric()''')
+  *
+  * - [[org.apache.flink.ml.nn.KNN.Blocks]]
+  * Sets the number of blocks into which the input data will be split. This number should be set
+  * at least to the degree of parallelism. If no value is specified, then the parallelism of the
+  * input [[DataSet]] is used as the number of blocks. (Default value: '''None''')
+  *
+  * - [[org.apache.flink.ml.nn.KNN.UseQuadTreeParam]]
+  * A boolean variable that whether or not to use a Quadtree to partition the training set
+  * to potentially simplify the KNN search.  If no value is specified, the code will
+  * automatically decide whether or not to use a Quadtree.  Use of a Quadtree scales well
+  * with the number of training and testing points, though poorly with the dimension.
+  * (Default value:  ```None```)
+  *
+  * - [[org.apache.flink.ml.nn.KNN.SizeHint]]
+  * Specifies whether the training set or test set is small to optimize the cross
+  * product operation needed for the KNN search.  If the training set is small
+  * this should be `CrossHint.FIRST_IS_SMALL` and set to `CrossHint.SECOND_IS_SMALL`
+  * if the test set is small.
+  * (Default value:  ```None```)
+  *
+  */
+
+class KNN extends Predictor[KNN] {
+
+  import KNN._
+
+  var trainingSet: Option[DataSet[Block[FlinkVector]]] = None
+
+  /** Sets K
+    * @param k the number of selected points as neighbors
+    */
+  def setK(k: Int): KNN = {
+    require(k > 0, "K must be positive.")
+    parameters.add(K, k)
+    this
+  }
+
+  /** Sets the distance metric
+    * @param metric the distance metric to calculate distance between two points
+    */
+  def setDistanceMetric(metric: DistanceMetric): KNN = {
+    parameters.add(DistanceMetric, metric)
+    this
+  }
+
+  /** Sets the number of data blocks/partitions
+    * @param n the number of data blocks
+    */
+  def setBlocks(n: Int): KNN = {
+    require(n > 0, "Number of blocks must be positive.")
+    parameters.add(Blocks, n)
+    this
+  }
+
+  /**
+   * Sets the Boolean variable that decides whether to use the QuadTree or not
+   */
+  def setUseQuadTree(useQuadTree: Boolean): KNN = {
+    if (useQuadTree){
+      require(parameters(DistanceMetric).isInstanceOf[SquaredEuclideanDistanceMetric] ||
+        parameters(DistanceMetric).isInstanceOf[EuclideanDistanceMetric])
```

Comment:
In `minDist`, an exception will be thrown.  I put this here in case the quadtree is used outside of knn.  This is also the only place where I need the assumption about being Euclidean (more details on this in your question below)